osu.Game/Overlays/Comments/CommentsPage.cs

```diff
@@ -52,18 +62,60 @@ private void load(OverlayColourProvider colourProvider)
                 return;
             }
 
-            foreach (var c in commentBundle.Comments)
+            commentBundle.Comments.ForEach(c =>
             {
                 if (c.IsTopLevel)
-                {
-                    flow.Add(new DrawableComment(c)
-                    {
-                        ShowDeleted = { BindTarget = ShowDeleted }
-                    });
-                }
+                    flow.Add(createCommentWithReplies(c, commentBundle));
+            });
+        }
+
+        private DrawableComment createCommentWithReplies(Comment comment, CommentBundle commentBundle)
+        {
+            var drawableComment = createDrawableComment(comment);
+
+            var replies = commentBundle.Comments.Where(c => c.ParentId == comment.Id);
+
+            if (replies.Any())
+            {
+                replies.ForEach(c => c.ParentComment = comment);
+                drawableComment.OnLoadComplete += _ => drawableComment.AddReplies(replies.Select(reply => createCommentWithReplies(reply, commentBundle)));
```

Comment:
IMO this is *very* dangerous and insidious recursion. It's nice in concept, but I'd be incredibly hesitant to let this be.

Such recursion in `OnLoadComplete` introduces a branching factor. Say you have a comment tree, with 1 parent comment, which is 5 levels deep, and each parent comment has 3 child comments. That tree then has (3<sup>6</sup> - 1)/2 = 364 comments, which is not that much.

However, for each child comment you're launching a linear scan of `commentBundle`, which'd mean 364 linear scans of the entire comment set. Moreover, I'd have to check how the threading with `OnLoadComplete` is done, but if it's done using worker threads on the thread pool, you could end up saturating the pool at the lower levels of the tree (since the further down you go, the more times the recursion fires). Level 0 of the tree only has the root node, level 1 - 3 nodes, level 2 - 9, level 3 - 27, and so on up to level 5 = 243 nodes. Or, if it's done on the draw thread, you'll starve that threads with `OnLoadComplete` callbacks.

Reply:
In theory I wanted to not use `OnLoadComplete` at all, but we can't use `AddReplies` if the drawable isn't loaded since it has `LoadAsync` function.

What about scanning `commentBundle` - it would be nice if all the comments were already ordered in tree-like manner. But may be they already are - I don't know yet (will look at it). Anyway I'll revisit all that system

Reply:
Also, is it worth to work on? I mean, is handling all that logic in `CommentsPage` (excluding this case) is good enough?

Reply:
Just a side note: we receiving maximum 50 comments at once (thanks to web api), so may be it's not a big deal?
(I'll remove `LoadComplete` anyway, already have some thoughts)